akka-typed/src/main/scala/akka/typed/Sessions.scala

```diff
@@ -0,0 +1,332 @@
+
+    /**
+     * Evaluate the predicate for this process’ result and abort if it yields `false`.
+     */
+    def filter(pred: T => Boolean): Process[T] = flatMap(t => if (pred(t)) unit(t) else fail)
+
+    /**
+     * Evaluate the predicate for this process’ result and abort if it yields `false`.
+     */
+    def withFilter(pred: T => Boolean): Process[T] = filter(pred)
+
+    /**
+     * Filter and transform this process.
+     */
+    def collect[U](pf: PartialFunction[T, U]): Process[U] = flatMap(t => if (pf.isDefinedAt(t)) unit(pf(t)) else fail)
+
+    /**
+     * Yields both values after both processes terminate; both processes run
+     * concurrently.
+     */
+    def join[U](other: Process[U]): Process[(T, U)]
+
+    /**
+     * Yields the value of the first process to terminate, canceling the other;
+     * both processes run concurrently.
+     */
+    def race[U >: T](other: Process[U]): Process[U]
+  }
+
+  sealed trait Termination
+  case object Halted extends Termination
+  case object Canceled extends Termination
+  case object Failed extends Termination
+
+  /**
+   * Ensures that onTerminate runs when the given process terminates.
+   * It should be noted that onTerminate can only perform synchronous side-effects.
+   */
+  def trap[T](onTerminate: Termination => Unit)(p: => Process[T]): Process[T] =
+    try Trap(p.sorry, onTerminate)
+    catch {
+      case NonFatal(ex) =>
+        onTerminate(Failed)
+        FailedImpl
+    }
+
+  /**
+   * A Channel can be sent to (it is an ActorRef) and read from (using the `read` process).
+   */
+  trait Channel[T] { this: ChannelImpl[T] =>
+    /**
+     * Obtain a reference that allows sending to this channel.
+     */
+    def ref: ActorRef[T]
+    /**
+     * Release all resources associated with this channel, future reads will fail.
+     */
+    def seal(): Unit
+  }
+
+  /**
+   * Creates a fresh channel with the given buffer capacity.
+   */
+  def channel[T](buffer: Int): Channel[T] = new ChannelImpl(buffer, actorContext)
+
+  /**
+   * Lifts a value into a process.
+   */
+  def unit[T](value: T): Process[T] = Value(value)
+
+  /**
+   * Spawns that other process to run concurrently, without awaiting its result.
+   */
+  def fork(other: Process[Any]): Process[Unit] = Fork(other.sorry)
+
+  /**
+   * Reads from a channel.
+   */
+  def read[T](c: Channel[T]): Process[T] = Read(c.sorry, false)
+
+  /**
+   * Reads one message from the channel and then seals it.
+   */
+  def readAndSeal[T](c: Channel[T]): Process[T] = Read(c.sorry, true)
+
+  /**
+   * Produces the given value after the given delay.
+   */
+  def timer[T](value: T, delay: FiniteDuration): Process[T] = Schedule(delay, Value(value))
+
+  /**
+   * Create a process that fails after the given delay.
+   */
+  def timeout(delay: FiniteDuration): Process[Nothing] = Schedule(delay, FailedImpl)
+
+  /**
+   * Create a process that signals termination.
+   */
+  def halt: Process[Nothing] = HaltedImpl
```

Comment:
No, after `halt` there cannot be anything, in other words this process yields nothing. There is a combinator missing, BTW, which allows recovery. I have not yet thought through all the consequences this would entail.